core/src/main/java/com/statful/client/core/buffer/AggregatedBuffer.java

```diff
@@ -0,0 +1,172 @@
+            }
+        }
+
+        aggregatedBuffer.put(aggregationFrequency.toString(), aggregatedFreqBuffer);
+        buffer.put(aggregation.toString(), aggregatedBuffer);
+
+        return aggregatedFreqBuffer.offer(metric);
+    }
+
+    /**
+     * Reads the buffer contents for a particular aggregation.
+     * @param aggregation The {@link Aggregation} aggregation to inspect the buffer
+     * @param aggregationFrequency The {@link AggregationFrequency} aggregation frequency to inspect the buffer
+     * @return A {@link String} with all the metrics for a particular aggregation
+     */
+    public final String readBuffer(final Aggregation aggregation, final AggregationFrequency aggregationFrequency) {
+        Map<String, ArrayBlockingQueue<String>> aggregatedBuffer = buffer.get(aggregation.toString());
+        ArrayBlockingQueue<String> aggregatedFreqBuffer;
+
+        if (aggregatedBuffer != null) {
+            aggregatedFreqBuffer = aggregatedBuffer.get(aggregationFrequency.toString());
+
+            if (aggregatedFreqBuffer != null) {
+                Collection<String> messages = new ArrayList<String>();
+                aggregatedFreqBuffer.drainTo(messages, flushSize);
+
+                StringBuilder sb = new StringBuilder();
+                for (String metric : messages) {
+                    sb.append(metric).append("\n");
+                }
+
+                return sb.toString();
+            }
+        }
+
+        return "";
+    }
+
+    /**
+     * Return the current aggregations buffers.
+     * @return A {@link Set} set with the current aggregations buffers
+     */
+    public final Set<Aggregation> getAggregations() {
+        Set<String> aggregationNames = buffer.keySet();
+        Set<Aggregation> aggregations = new HashSet<Aggregation>();
+
+        for (String aggregation : aggregationNames) {
+            aggregations.add(Aggregation.valueOf(aggregation));
+        }
+
+        return aggregations;
+    }
+
+    /**
+     * Return the current aggregation frequencies buffers.
+     * @param aggregation The {@link Aggregation} aggregation to inspect the buffer
+     * @return A {@link Set} set with the current aggregations frequencies buffers
+     */
+    public final Set<AggregationFrequency> getAggregationFrequencies(final Aggregation aggregation) {
+        Map<String, ArrayBlockingQueue<String>> aggregationFreqBuffer = buffer.get(aggregation.toString());
+
+        Set<AggregationFrequency> aggregationFrequencies = new HashSet<AggregationFrequency>();
+
+        if (aggregationFreqBuffer != null) {
+            for (String aggregationFreq : aggregationFreqBuffer.keySet()) {
+                aggregationFrequencies.add(AggregationFrequency.valueOf(aggregationFreq));
+            }
+        }
+
+        return aggregationFrequencies;
+    }
+
+    @Override
+    public final boolean isTimeToFlush() {
+        Set<Aggregation> aggregations = getAggregations();
```

Comment:
I don't like the current implementation, I've mentioned that on the code review.

With that being said I don't like prefixing meta info into the metric data as that would require loads of string manipulation and could potentially mess up the metric data since it would require logic both when writing/reading from the buffer. 

In the meantime I thought on having a composite key class that would have two fields: `Aggregation` and `AggregationFrequency` and would serialize to and from a concatenated `String` which would work as a buffer map key.

That would effectively make the complexity linear for buffer reads since the aggregated buffer would be a single flat map and `isTimeToFlush` would be `O(N)` with `N` being the max number of `Aggregation/AggregationFrequencies` combinations which we know before hand is relatively small.

I'm thinking of refactoring this on the next iteration and merging as it is right now. Thoughts?